src/digest.h

```diff
@@ -0,0 +1,90 @@
+/*
+
+  digest -- hash digest functions for R
+
+  Copyright (C) 2003 - 2024  Dirk Eddelbuettel <edd@debian.org>
+
+  This file is part of digest.
+
+  digest is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  digest is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with digest.  If not, see <http://www.gnu.org/licenses/>.
+
+*/
+
+#include <stdint.h> // for uint32_t
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <R.h>
+#include <Rdefines.h>
+#include <Rinternals.h>
+#include <Rversion.h>
+#include <inttypes.h>
+
+#ifdef _WIN32
+#include <Windows.h>
+#endif
+
+// Also already used in sha2.h
+//
+// We can rely on WORDS_BIGENDIAN only be defined on big endian systems thanks to Rconfig.
+//
+// A number of other #define based tests are in other source files here for different hash
+// algorithm implementations notably crc32c, pmurhash, sha2 and xxhash
+//
+// A small and elegant test is also in package qs based on https://stackoverflow.com/a/1001373
+
+// edd 02 Dec 2013  use Rconfig.h to define BYTE_ORDER, unless already defined
+#ifndef BYTE_ORDER
+// see sha2.c comments, and on the internet at large
+#define LITTLE_ENDIAN 1234
+#define BIG_ENDIAN    4321
+#ifdef WORDS_BIGENDIAN
+#define BYTE_ORDER  BIG_ENDIAN
+#else
+#define BYTE_ORDER  LITTLE_ENDIAN
+#endif
+#endif
+
+SEXP is_big_endian(void);
+SEXP is_little_endian(void);
+
+#if defined(R_VERSION) && R_VERSION >= R_Version(3,0,0)
+#define RVLENGTH(ROBJ) XLENGTH(ROBJ)
+#else
+#define RVLENGTH(ROBJ) LENGTH(ROBJ)
+#endif
```

Comment:
Nobody will compile on R older than 3.0.0.  We could just make it an error on config.

And then stick with `XLENGTH` which is known.

Reply:
works for me. what's the cleanest way to do that? seems like something to just specify in DESCRIPTION, right?

Reply:
Yes.  

Find any comparison for R_VERSION that are that ancient (as here R 3.0.0), default to the 'newer' case and set then in DESCRIPTION set 'R (>= 3.0.0)'.   [ Goes checking. ] Looks like that was the only one in src/.